### PR TITLE
Handle large numbers and -inf in losses

### PR DESCRIPTION
**Pull request by jvlmdr**
Adjusts order of operations to avoid precision issues (#243)

### CODE REVIEW

File: jaxopt/_src/loss.py [hunk 1]
```diff
@@ -57,7 +57,9 @@ def binary_logistic_loss(label: int, logit: float) -> float:
   # Softplus is the Fenchel conjugate of the Fermi-Dirac negentropy on [0, 1].
   # softplus = proba * logit - xlogx(proba) - xlogx(1 - proba),
   # where xlogx(proba) = proba * log(proba).
-  return softplus(logit) - label * logit
```
Comment (mblondel) [inline jaxopt/_src/loss.py:60]: Could you add this implementation to the tests and check that the new one recovers the same value?

File: jaxopt/_src/loss.py [hunk 2]
```diff
@@ -93,10 +96,9 @@ def multiclass_sparsemax_loss(label: int, scores: jnp.ndarray) -> float:
...
-  n_classes = scores.shape[0]
-  one_hot = jax.nn.one_hot(label, n_classes)
   proba = projection_simplex(scores)
   # Fenchel conjugate of the Gini negentropy, defined by:
   # cumulant = jnp.dot(proba, scores) + 0.5 * jnp.dot(proba, (1 - proba)).
-  cumulant = 0.5 + jnp.dot(proba , scores - 0.5 * proba)
-  return cumulant - jnp.dot(scores, one_hot)
+  scores = (scores - scores[label]).at[label].set(0.0)
+  return (jnp.dot(proba, jnp.where(proba, scores, 0.0))
+          + 0.5 * (1.0 - jnp.dot(proba, proba)))
```
Comment (mblondel) [inline jaxopt/_src/loss.py:104]: Same here, could you check in the tests against the old implementation.

**Review by mblondel — COMMENTED** [review]
Thanks for tackling all these edge cases. This is great!

Before we merge, could you add tests to check against the old implementation of the logistic and sparsemax losses? The old implementations are less robust but more readable so this is a great want to check correctness. The current tests are just sanity checks but they don't really ensure that the implementation is correct.

**Comment by jvlmdr** [comment]
Thanks for reviewing this! I have added those tests.

I thought I would mention that, since this implementation uses `ndarray.at`, it will only work when called on `jnp.ndarray` and not `onp.ndarray` types.

I could remove this, I just thought it was nicer for the purpose of gradient computation to force the difference to be zero, although it should be fine in practice. (It also enables the loss functions to support `+inf` for the positive label (avoiding `inf - inf => nan`), although this doesn't seem like an important use case. I didn't add a test for this because it would have complicated the implementation of the simplex projection to make the `sparsemax` loss pass.)

**Comment by mblondel** [comment]
Can't you just wrap the array with `logits = jnp.array(logits)` at the beginning?

**Comment by jvlmdr** [comment]
Yep, done. I left the typing annotation as `logits: jnp.ndarray`

**Comment by mblondel** [comment]
This is ready for merge thanks a lot! Could you squash commits? Thanks in advance!

**Comment by jvlmdr** [comment]
Yep, done! Thank you!
